### DISCUSSION TITLE
Tracking attribute changes over relations for compund objects

### DISCUSSION THREAD

**Discussion by tokarenko** [sqlalchemy/sqlalchemy#10616] [Usage Questions] [answered]
Please advise if it is possible to use [Mutation tracking](https://docs.sqlalchemy.org/en/20/orm/extensions/mutable.html) to detect changes in a attribute (of compound object) that is a relation to another object? In my particular case (see code below), I am trying to inspect 'CollectionPolicy' for changes in 'collection_objects' relation to 'CollectionObject'. I found advice that it may be done [via SQLAlchemy event subsystem](https://docs.sqlalchemy.org/en/20/_modules/examples/custom_attributes/listen_for_events.html). I am trying to explore SQLAlchemy Mutation tracking and Inspection engine for the purpose.

```
@mapper_registry.mapped
@dataclass
class CollectionObject:
    id: int = field(init=False)
    name: str
    description: str

    collection_policy_id: int = field(init=False)

    __table__ = Table(
        "collection_object",
        mapper_registry.metadata,
        Column("id", Integer, primary_key=True),
        Column("name", Text),
        Column("description", Text),
        Column("collection_policy_id", Integer, ForeignKey("collection_policy.id")),
    )

@mapper_registry.mapped
@dataclass
class CollectionPolicy:
    id: int = field(init=False)
    collection_objects: List[CollectionObject]

    __table__ = Table(
        "collection_policy",
        mapper_registry.metadata,
        Column("id", Integer, primary_key=True),
    )

    __mapper_args__ = {
        "properties": {
            "collection_objects": relationship("CollectionObject"),
        }
      }

engine = create_engine("postgresql+psycopg://user:password@server/database")
session_factory = sessionmaker(bind=engine)
with session_factory() as session
    policy = session.get(policy)
    policy.collection_objects[0].name = "newname"
    policy_state = inspect(policy)
    attribute_state = instance_state.attrs["collection_objects"]
    if attribute_state.history.has_changes():
        ...

```

**Comment by zzzeek** [comment]
the mutation tracking API is only used for immediate column-based attributes which refer to a composed Python object some kind; the most classical example is a JSON column that refers to a mutable Python structure like a dictionary.  It is not used for relationship mutation tracking.

in order to intercept changes on a relationship attribute such as `CollectionPolicy.collection_objects` you would be using the https://docs.sqlalchemy.org/en/20/orm/events.html#sqlalchemy.orm.AttributeEvents events.

**Reply by tokarenko** [reply]
Thank you, Mike, for your quick reply. As there will be many attributes to track, I have further question. May I somehow capture a change at instance level (without details)? I am not interested in capturing the details of the change just to signal to a compound object root that its part over a relation has changed. Maybe check in session.dirty for the related parts of compund object?

**Reply by zzzeek** [reply, marked as answer]
the attribute events are what fire off when the change actually happens.  Alternatively, if you want to scan for things that have changed within a flush event, that is, not grab the change as it happens, but just look for changes when it's time to talk to the database, you can yes look in session.dirty to see what objects changed, and if you scan those you can look into [InstanceState.attrs](https://docs.sqlalchemy.org/en/20/orm/internals.html#sqlalchemy.orm.InstanceState.attrs) for each object to see what objects were added/removed from collections.
